### PR TITLE
Neutral Diffusion: Add option for discontinuous reconstruction of T/S

### PR DESCRIPTION
**Pull request by ashao**
This PR adds the option to connect neutral surfaces using a discontinuous reconstruction of temperature and salinity (after drawing 100s of arrows and surfaces on paper, writing new logic, throwing it away, drawing more arrows, and finally coming up with a consistent algorithm). A notebook demonstrating the algorithm for some simple test cases can be found at: https://github.com/ashao/MOM6_analysis/blob/master/neutral_diffusion/neutral_diffusion_examples.ipynb. This is controlled by the runtime parameter NDIFF_CONTINUOUS which is by default True if using the original continuous parabolic reconstructions and false if doing a discontinuous PPM reconstruction.

The main deficiency of this algorithm as it stands is that T/S in the top and bottom boundary layers are using a piecewise continuous reconstruction. This leads to some odd behavior in the two bottom-most and topmost layers in the ocean_only/tracer_mixing experiment. If you have two columns with a density profile rho(z) and the other with rho(z)+epsilon, the entire top or bottom layer will be skipped whereas in actuality some part of the top and bottom layers should always overlap. However, in the interior the neutral surfaces seem to behave as expected.

Before finalizing this update to the neutral diffusion code, two more improvements could be made:
-Boundary extrapolation (to address the above deficiency) applied to both the T/S reconstructions but also the tracer being diffused. 
-Something other than a linear interpolation could be used to find where the neutral surface connects within a layer.

### CODE REVIEW

File: src/tracer/MOM_neutral_diffusion.F90 [hunk 1]
```diff
@@ -1102,7 +1472,7 @@ logical function neutral_diffusion_unit_tests(verbose)
   neutral_diffusion_unit_tests = neutral_diffusion_unit_tests .or. test_ifndp(v,-2.0, .5,  5.0, 0.5, 0.5, 'Check dP=0')
 
   ! Identical columns
-  call find_neutral_surface_positions(3, &
+  call find_neutral_surface_positions_continuous(3, &
```
Comment (adcroft) [inline src/tracer/MOM_neutral_diffusion.F90]: We'll need units tests for each or your _discontinuous functions.

File: src/tracer/MOM_neutral_diffusion.F90 [hunk 2]
```diff
@@ -877,16 +946,292 @@ subroutine find_neutral_surface_positions(nk, Pl, Tl, Sl, dRdTl, dRdSl, Pr, Tr,
...
     endif
 
   enddo neutral_surfaces
 
-end subroutine find_neutral_surface_positions
+end subroutine find_neutral_surface_positions_continuous
+
+!> Higher order version of find_neutral_surface_positions. Returns positions within left/right columns
+!! of combined interfaces using intracell reconstructions of T/S
+subroutine find_neutral_surface_positions_discontinuous(nk, Pres_l, Tl, Sl, dRdT_l, dRdS_l, &
```
Comment (adcroft) [inline src/tracer/MOM_neutral_diffusion.F90]: It's a shame we can't reuse the same algorithm, or combine them. Debugging these sorts of algorithms is painful. Is this the same algorithm but with different data structures being passed in?

One test to try is to set the discontinuous edge values to the interface values of the "continuous" approach. It should give the same solution as the continuous but with 2*G%ke vanished sub-layers.

File: src/tracer/MOM_neutral_diffusion.F90 [hunk 3]
```diff
@@ -266,44 +279,107 @@ subroutine neutral_diffusion_calc_coeffs(G, GV, h, T, S, EOS, CS)
...
+        endif
+      enddo
+    else ! Discontinuous reconstruction
+      do k = 1, G%ke
+        call calculate_density_derivs(T_i(:,j,k,1), S_i(:,j,k,1), Pint(:,j,k), &
+                                      dRdT_i(:,j,k,1), dRdS_i(:,j,k,1), G%isc-1, G%iec-G%isc+3, EOS)
+        if (k<=G%ke) then
+          Pint(:,j,k+1) = Pint(:,j,k) + h(:,j,k) * GV%H_to_Pa ! Pressure at next interface, k+1 (Pa)
+          call calculate_density_derivs(T_i(:,j,k,2), S_i(:,j,k,2), Pint(:,j,k+1), &
+                                        dRdT_i(:,j,k,1), dRdS_i(:,j,k,2), G%isc-1, G%iec-G%isc+3, EOS)
```
Comment (adcroft) [inline src/tracer/MOM_neutral_diffusion.F90]: Bug: there's a `1` here that should be a `2`.

Comment (adcroft) [inline src/tracer/MOM_neutral_diffusion.F90]: PS> For the record, the use of indices 1,2 for left,right in the reconstruction code is not my doing and I don't like it because you can't read the code. One day we'll clean up the reconstruction interfaces...

File: src/tracer/MOM_neutral_diffusion.F90 [hunk 4]
```diff
@@ -266,44 +279,107 @@ subroutine neutral_diffusion_calc_coeffs(G, GV, h, T, S, EOS, CS)
...
+  real, dimension(:,:,:,:), allocatable        :: S_i     ! Top edge reconstruction of salinity (ppt)
+  real, dimension(:,:,:,:), allocatable        :: dRdT_i     ! dRho/dT (kg/m3/degC) at top edge
+  real, dimension(:,:,:,:), allocatable        :: dRdS_i     ! dRho/dS (kg/m3/ppt) at top edge
+  real, dimension(:,:),     allocatable        :: ppoly_coefficients ! PPM coefficients
+  ! Allcoatable variables for continuous reconstructions
+  real, dimension(:,:,:), allocatable        :: dRdT ! dRho/dT (kg/m3/degC) at interface
+  real, dimension(:,:,:), allocatable        :: dRdS ! dRho/dS (kg/m3/ppt) at interfaces
+
+  if (CS%continuous_reconstruction) then
+    if (.not. ALLOCATED(dRdT)) ALLOCATE(dRdT(SZI_(G),SZJ_(G),SZK_(G)+1))
```
Comment (ashao) [inline src/tracer/MOM_neutral_diffusion.F90]: Whoops, this was an oversight before making the PR. I think when I wrote it I was trying to be super-safe that there was no memory leak.

File: src/tracer/MOM_neutral_diffusion.F90 [hunk 5]
```diff
@@ -266,44 +279,107 @@ subroutine neutral_diffusion_calc_coeffs(G, GV, h, T, S, EOS, CS)
...
+        call interface_scalar(G%ke, h(i,j,:), T(i,j,:), Tint(i,j,:), 2)
+        call interface_scalar(G%ke, h(i,j,:), S(i,j,:), Sint(i,j,:), 2)
+      else
+        call PPM_reconstruction(G%ke, h(i,j,:), T(i,j,:), T_i(i,j,:,:), ppoly_coefficients)
+        if (CS%boundary_extrapolation) then
+          call PPM_boundary_extrapolation(G%ke, h(i,j,:), T(i,j,:), T_i(i,j,:,:), ppoly_coefficients)
+        endif
+        call PPM_reconstruction(G%ke, h(i,j,:), S(i,j,:), S_i(i,j,:,:), ppoly_coefficients)
+        if (CS%boundary_extrapolation) then
+          call PPM_boundary_extrapolation(G%ke, h(i,j,:), S(i,j,:), S_i(i,j,:,:), ppoly_coefficients)
```
Comment (ashao) [inline src/tracer/MOM_neutral_diffusion.F90]: That's a good point that I hadn't thought of. My main reason for wanting to do some kind of extrapolation is that it seems somewhat unphysical that you would have a large jump in the sum of hEff in the case between two columns with identical columns versus the case where one is only very slightly lighter or denser than the other because you exclude the entirety of the top and bottom cells.

File: src/tracer/MOM_neutral_diffusion.F90 (hunk 3 above)
Comment (ashao) [inline src/tracer/MOM_neutral_diffusion.F90]: Thanks for the catch. Also, I agree that it's a little bit difficult for 1,2 to map onto left right edge values. At the same time the solution that I can think of off the top I my head is to have separate variables for the two of them which can lead to a huge amount of input arguments (which was the original I had this algorithm coded).

File: src/tracer/MOM_neutral_diffusion.F90 [hunk 6]
```diff
@@ -357,36 +433,28 @@ subroutine neutral_diffusion(G, GV, h, Coef_x, Coef_y, Tracer, m, dt, name, CS)
 
   ! x-flux
   do j = G%jsc,G%jec ; do I = G%isc-1,G%iec
-    if (G%mask2dCu(I,j)>0.) then
-      call neutral_surface_flux(nk, h(i,j,:), h(i+1,j,:),       &
+      call neutral_surface_flux(nk, CS%nsurf, h(i,j,:), h(i+1,j,:),       &
```
Comment (ashao) [inline src/tracer/MOM_neutral_diffusion.F90]: Mistake, I'll add it back in.

File: src/tracer/MOM_neutral_diffusion.F90 (hunk 2 above)
Comment (ashao) [inline src/tracer/MOM_neutral_diffusion.F90]: I spent a bit of time trying to shoehorn your original algorithm into the discontinuous one, but to no avail. What it boiled down to was that the original algorithm relies on the fact that the layer index always corresponds to the top interface. With the discontinuous reconstructions a different mapping between interface and layer exists, plus you may potentially need to search across the discontinuity.

I did do something similar to the test you suggested when I was prototyping the logic for the new one for some really simple cases that produced qualitatively similar results (didn't check for bitwise reproduction). If they do, then I agree that the code could be collapsed so that there's not a huge bit of similar code.

File: src/tracer/MOM_neutral_diffusion.F90 (hunk 1 above)
Comment (ashao) [inline src/tracer/MOM_neutral_diffusion.F90]: As you noted some time before, the discontinuous reconstructions don't always lend themselves nicely to whole number results. That being said, it's on my to-do list before submitting an actual PR.

**Comment by ashao** [comment]
Many thanks for the review. I'll close this PR for now and reopen when it's ready to be looked at again.

**Comment by ashao** [comment]
The new neutral diffusion code is ready to be pulled into the main code. For now, the original reconstruction code is retained and remains the default so answers should not change. The improvements made here (discontinuous reconstruction and iterating to find a true neutral position) increases the computational cost of tracer diffusion by a factor of 3-4 in OM4_05, about the same cost as the advection part of the routine. Unit tests for a wide variety of cases are included.

**Comment by Hallberg-NOAA** [comment]
There are a couple of the density derivative routines that are being duplicated for scalar and array arguments, including some with messy calculations.  I would prefer that the pattern of calculate_density be followed, in that the scalar versions are not coded separately but rather copy their arguments into 1-point arrays and then call the array version. Doing it that way instead would avoid having two separate (and potentially inconsistent) implementations of the algorithms.

**Comment by Hallberg-NOAA** [comment]
The lengthy list of commits in this PR appears to change both answers and (less concerningly) the parameter_doc files in several test cases.  However, I do not see anything in the commit logs that indicate which commits actually change the answers.  If there are answer-changing modifications, please break up this pull request into the larger part that does not change answers and a very limited set of code changes that do change answers.

**Comment by ashao** [comment]
None of the commits should have changed answers and the cases checked by Travis passed. To help narrow down where I may have inadvertently introduced a change, could you send a subset of the test cases which failed their checks?

**Comment by ashao** [comment]
Oh wait, I see changes that shouldn't have been there. Submitting yet another commit shortly
